src/components/functional_module/src/plugin_manager.cc

```diff
@@ -0,0 +1,410 @@
+  std::vector<std::string> plugin_files = file_system::ListFiles(plugin_path);
+  for (size_t i = 0; i < plugin_files.size(); ++i) {
+    size_t pos = plugin_files[i].find_last_of(".");
+    if (std::string::npos != pos) {
+      if (plugin_files[i].substr(pos + 1).compare("so") != 0) {
+        continue;
+      }
+    } else {
+      continue;
+    }
+    std::string full_name = plugin_path + '/' + plugin_files[i];
+    void* generic_plugin_dll = dlopen(full_name.c_str(), RTLD_LAZY);
+    if (NULL == generic_plugin_dll) {
+      LOG4CXX_ERROR(logger_,
+                    "Failed to open dll " << plugin_files[i] << "\n"
+                                          << dlerror());
+      continue;
+    }
+    typedef GenericModule* (*Create)();
+    Create create_manager =
+        reinterpret_cast<Create>(dlsym(generic_plugin_dll, "Create"));
+    char* error_string = dlerror();
+    if (NULL != error_string) {
+      LOG4CXX_ERROR(logger_,
+                    "Failed to export dll's " << plugin_files[i] << " symbols\n"
+                                              << error_string);
+      dlclose(generic_plugin_dll);
+      continue;
+    }
+    ModulePtr module = create_manager();
+    if (!module) {
+      LOG4CXX_ERROR(logger_,
+                    "Failed to create plugin main class " << plugin_files[i]);
+      dlclose(generic_plugin_dll);
+      continue;
+    } else {
+      LOG4CXX_DEBUG(logger_,
+                    "Opened and working plugin from " << plugin_files[i]
+                                                      << " with id "
+                                                      << module->GetModuleID());
+      dlls_.insert(std::pair<ModuleID, void*>(module->GetModuleID(),
+                                              generic_plugin_dll));
+      plugins_.insert(
+          std::pair<ModuleID, ModulePtr>(module->GetModuleID(), module));
+      std::deque<RCFunctionID> subscribers =
+          module->GetPluginInfo().rc_function_list;
+      for (size_t i = 0; i < subscribers.size(); ++i) {
+        mobile_subscribers_.insert(
+            std::pair<RCFunctionID, ModulePtr>(subscribers[i], module));
+      }
+
+      std::deque<HMIFunctionID> hmi_subscribers =
+          module->GetPluginInfo().hmi_function_list;
+      for (size_t i = 0; i < hmi_subscribers.size(); ++i) {
+        hmi_subscribers_.insert(
+            std::pair<HMIFunctionID, ModulePtr>(hmi_subscribers[i], module));
+      }
+      module->set_service(service_);
+      module->AddObserver(this);
+    }
+  }
+  return plugins_.size();
+}
+
+void PluginManager::UnloadPlugins() {
+  for (Modules::iterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    it->second->RemoveObserver(this);
+  }
+  plugins_.clear();
+
+  for (std::map<ModuleID, void*>::iterator it = dlls_.begin();
+       dlls_.end() != it;
+       ++it) {
+    dlclose(it->second);
+  }
+  dlls_.clear();
+}
+
+// TODO(VS): Optimize similar code in ProcessMessage, IsMessageForPlugin,
```

Comment:
does TODO need to be included in this RC PR?

Reply:
@LuxoftAKutsan please do so

Reply:
@jacobkeeler not actual TODOs removed